### PR TITLE
feat(eslint-plugin): [prefer-promise-reject-errors] add rule

### LINKED ISSUE #7687: Base rule extension: [prefer-promise-reject-errors]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+base+rule+extension%22) and found none that match my proposal.
- [X] I have searched the [current extension rule list](https://typescript-eslint.io/rules/#extension-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the base rule

https://eslint.org/docs/latest/rules/prefer-promise-reject-errors

### Description

the `prefer-promise-reject-errors` rule should be extended as the base rule does not catch cases where the value is not a literal

### Fail

```typescript
const foo = "asdf"
Promise.reject(foo) // no error
```

### Pass

```typescript
const foo = new Error("asdf")
Promise.reject(foo) // no error
```

### Additional Info

originally raised #7673 and https://github.com/eslint/eslint/issues/17592 for `no-throw-literal` because i wasn't aware there was already a base rule for this

### PR DESCRIPTION
**Pull request by auvred**
## PR Checklist

- [x] Addresses an existing open issue: fixes #7687
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

## Overview

I tried to migrate the [original rule](https://github.com/eslint/eslint/blob/main/lib/rules/prefer-promise-reject-errors.js) with only type information added (it allows to detect `PromiseConstructor` like and `Error` like variables)

Also I moved `isErrorLike` function to the `type-utils` package (I hope it's the right place for this function) as its logic may be reused both in `prefer-promise-reject-errors` and `no-throw-literal`

I copied incorrect and correct examples from [`no-throw-literal`](https://typescript-eslint.io/rules/no-throw-literal) because I found them applicable. But maybe there are too many examples for this rule

### CODE REVIEW

File: packages/eslint-plugin/tests/rules/prefer-promise-reject-errors.test.ts
Comment (StyleShit) [inline packages/eslint-plugin/tests/rules/prefer-promise-reject-errors.test.ts:84]: those 2 are not valid code regardless of the rule

File: packages/eslint-plugin/tests/rules/prefer-promise-reject-errors.test.ts
Comment (StyleShit) [inline packages/eslint-plugin/tests/rules/prefer-promise-reject-errors.test.ts:159]: maybe also add a passing case for `reject(true && new Error())`?

File: packages/eslint-plugin/tests/rules/prefer-promise-reject-errors.test.ts
Comment (StyleShit) [inline packages/eslint-plugin/tests/rules/prefer-promise-reject-errors.test.ts:193]: same as above, maybe add a passing case with the same approach of renaming

File: packages/eslint-plugin/tests/rules/prefer-promise-reject-errors.test.ts
Comment (StyleShit) [inline packages/eslint-plugin/tests/rules/prefer-promise-reject-errors.test.ts:197]: hmm, WDYT about this one (in both valid and invalid cases)?
```TS
'new Promise((...args) => args[1](5))',
```

**Review by StyleShit — CHANGES_REQUESTED** [review]

File: packages/eslint-plugin/tests/rules/prefer-promise-reject-errors.test.ts
Comment (auvred) [reply packages/eslint-plugin/tests/rules/prefer-promise-reject-errors.test.ts:90]: I just added them to improve test coverage. This is a valid syntax, although it throws an exception in runtime

File: packages/eslint-plugin/tests/rules/prefer-promise-reject-errors.test.ts
Comment (auvred) [reply packages/eslint-plugin/tests/rules/prefer-promise-reject-errors.test.ts:203]: I thought about it, but I think we can do it in one of the following prs to make it easier to review the current one

Also, if we add support for this case, we would probably need to support a plenty of similar cases such as:

```tsx
new Promise((args) => { const [,reject] = args; reject(); })
```

```tsx
let reject
new Promise((resolve, _reject) => reject = _reject)
reject()
```

```tsx
new Promise(resolve, _reject) => {
  reject = _reject
  reject()
})
```
and so on...

File: packages/eslint-plugin/tests/rules/prefer-promise-reject-errors.test.ts
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/tests/rules/prefer-promise-reject-errors.test.ts:223]: [Testing] We've been moving away from auto-generated tests. They have a habit of being imprecise and hard to statically analyze. We'd also want to be able to customize/see more directly what line & column ranges the errors would be reported on. -1 from me on the `.map`, can you switch to direct object literals please?

File: packages/eslint-plugin/src/rules/prefer-promise-reject-errors.ts
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/prefer-promise-reject-errors.ts:135]: [Testing] I tried removing `!ref.isRead() ||` and none of the unit tests failed. Is there a reason to include it that can be tested?

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
Awesome, really solid work here - thanks for working on it @auvred!

It's interesting that this is the first rule to really need to check whether something is actually a `'PromiseConstructor'`. Others have just approximated comparisons to `Thenable`. 

A few touchups & refactors requested, but the general direction looks great to me! 🚀

File: packages/eslint-plugin/tests/rules/prefer-promise-reject-errors.test.ts
Comment (JoshuaKGoldberg) [reply packages/eslint-plugin/tests/rules/prefer-promise-reject-errors.test.ts:90]: No need to include these IMO, the rule doesn't really check for them and they'd give a runtime + TypeScript complaint.

File: packages/eslint-plugin/tests/rules/prefer-promise-reject-errors.test.ts
Comment (JoshuaKGoldberg) [reply packages/eslint-plugin/tests/rules/prefer-promise-reject-errors.test.ts:203]: Yeah I don't think we need to check these kinds of `...args` cases 🙂 that's rather more complex than what most rules try to handle.

File: packages/eslint-plugin/src/rules/prefer-promise-reject-errors.ts
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/prefer-promise-reject-errors.ts:26]: What do you think about recommending this in the `strict-type-checked` config? I think it'd be reasonable to as we already include https://typescript-eslint.io/rules/no-throw-literal in there.

File: packages/eslint-plugin/docs/rules/prefer-promise-reject-errors.md
Comment (auvred) [reply packages/eslint-plugin/docs/rules/prefer-promise-reject-errors.md:14]: Yep, thanks! Reduced it https://github.com/typescript-eslint/typescript-eslint/pull/8011/commits/5263fa80ed318f9ae7cf8abbb71c6b3dc91b529f

File: packages/eslint-plugin/docs/rules/prefer-promise-reject-errors.md
Comment (auvred) [reply packages/eslint-plugin/docs/rules/prefer-promise-reject-errors.md:32]: Yep, I implied that these comments would separate the sections. With less count of examples it looks cleaner. Please take a look again, I'm not sure if I should delete these comments

File: packages/eslint-plugin/src/rules/prefer-promise-reject-errors.ts
Comment (auvred) [reply packages/eslint-plugin/src/rules/prefer-promise-reject-errors.ts]: Not sure about the function name, but I extracted it: https://github.com/typescript-eslint/typescript-eslint/pull/8011/files#diff-75626015398ba4cc6293d6ec23077170a50f23e306b326a44966e44448c3edf2

---

TODO: refactor these places to use this function
https://github.com/typescript-eslint/typescript-eslint/blob/705370ac0d9c54081657b8855b398e57d6ea4ddb/packages/eslint-plugin/src/rules/no-implied-eval.ts#L82-L90 https://github.com/typescript-eslint/typescript-eslint/blob/705370ac0d9c54081657b8855b398e57d6ea4ddb/packages/eslint-plugin/src/rules/no-implied-eval.ts#L145-L153

File: packages/eslint-plugin/src/rules/prefer-promise-reject-errors.ts
Comment (auvred) [reply packages/eslint-plugin/src/rules/prefer-promise-reject-errors.ts]: Thx! https://github.com/typescript-eslint/typescript-eslint/pull/8011/commits/a680b8df2b2c5472e7fe7c7a639d6849b81188ce

File: packages/eslint-plugin/src/rules/prefer-promise-reject-errors.ts
Comment (auvred) [reply packages/eslint-plugin/src/rules/prefer-promise-reject-errors.ts]: You're right, thanks, check for `CallExpression` is already a check for `isRead`

File: packages/eslint-plugin/src/rules/prefer-promise-reject-errors.ts
Comment (auvred) [reply packages/eslint-plugin/src/rules/prefer-promise-reject-errors.ts:27]: Done https://github.com/typescript-eslint/typescript-eslint/pull/8011/commits/f2ee5d967496beda88be1e8c25feb317551a5097

File: packages/eslint-plugin/tests/rules/prefer-promise-reject-errors.test.ts
Comment (auvred) [reply packages/eslint-plugin/tests/rules/prefer-promise-reject-errors.test.ts:85]: Ah if we remove these 2 cases, the coverage is no longer 100% :(
100% coverage


This branch won't be tested: https://github.com/typescript-eslint/typescript-eslint/pull/8011/files#diff-801508874bd1d1cb9f122a9ca577fdf2350bfd756c5e63572487b65b859d6799R112-R115

But if we remove this branch, the rule will throw runtime error during linting `new Promise()`. I suppose it shouldn't break on valid syntax
Is `/* istanbul ignore */` applicable for such a case?

Comment (JoshuaKGoldberg) [reply packages/eslint-plugin/tests/rules/prefer-promise-reject-errors.test.ts:85]: `/* istanbul ignore */` is almost never appropriate (even most or all of our existing cases should probably be replaced with assertions or general-use utilities). 

If the rule would _crash_ in a particular case then I guess including the test makes sense yeah.

File: packages/eslint-plugin/tests/rules/prefer-promise-reject-errors.test.ts
Comment (auvred) [inline packages/eslint-plugin/tests/rules/prefer-promise-reject-errors.test.ts:790]: Thanks to @Tjstretchalot for pointing on this! https://github.com/typescript-eslint/typescript-eslint/pull/8094#issuecomment-1864771039

File: packages/eslint-plugin/src/rules/prefer-promise-reject-errors.ts
Comment (JoshuaKGoldberg) [reply packages/eslint-plugin/src/rules/prefer-promise-reject-errors.ts]: LGTM! If you want to do that refactor in this PR I'd happily approve. Or we can leave it as a `good first issue` followup. Either way.

File: packages/eslint-plugin/docs/rules/prefer-promise-reject-errors.md
Comment (JoshuaKGoldberg) [reply packages/eslint-plugin/docs/rules/prefer-promise-reject-errors.md:32]: One of the issues with setting a precedent of `//` comment separators is that a lot of rules end up having many different sections of responsibility. It's a slippery slope, if you'll pardon the known fallacy...

The `foo.bar` examples don't really speak to me. They don't introduce any new type concepts. They just show that the rule works in edge cases. What do you think about removing them and then removing section comments?

**Review by JoshuaKGoldberg — APPROVED** [review]
This looks lovely, thanks! 🙌 

No blocking requests (other than the _"guarantee"_ nit) from me. Just some touchup suggestions. Feel free to wontfix or do them, and then we can merge. Or if you don't have time I'll take another look soon. Cheers!

File: packages/eslint-plugin/docs/rules/prefer-promise-reject-errors.md
Comment (auvred) [reply packages/eslint-plugin/docs/rules/prefer-promise-reject-errors.md]: Thanks!

File: packages/type-utils/src/isBuiltinSymbolLike.ts
Comment (auvred) [reply packages/type-utils/src/isBuiltinSymbolLike.ts:1]: Makes perfect sense! Thank you

File: packages/eslint-plugin/src/rules/prefer-promise-reject-errors.ts
Comment (auvred) [reply packages/eslint-plugin/src/rules/prefer-promise-reject-errors.ts]: For now I just split it up into variables, if I can improve that, please tell me

File: packages/eslint-plugin/docs/rules/prefer-promise-reject-errors.md
Comment (auvred) [reply packages/eslint-plugin/docs/rules/prefer-promise-reject-errors.md]: Nice catch! +1 on asserting that examples under the *Correct* header don't contain any errors

> Perhaps we could also test that code under _**Incorrect**_ tab contains at least one lint error, and _**Correct**_ code contains no errors


https://github.com/typescript-eslint/typescript-eslint/issues/8134

File: packages/eslint-plugin/docs/rules/prefer-promise-reject-errors.md
Comment (auvred) [reply packages/eslint-plugin/docs/rules/prefer-promise-reject-errors.md]: Thanks for the explanation, I removed all of them: https://github.com/typescript-eslint/typescript-eslint/pull/8011/commits/eaab9eb552b9bc2b854625f286f7d0418062b27d

File: packages/eslint-plugin/src/rules/prefer-promise-reject-errors.ts
Comment (auvred) [reply packages/eslint-plugin/src/rules/prefer-promise-reject-errors.ts]: Let's open a followup with `good first issue` ;)

File: packages/eslint-plugin/src/rules/prefer-promise-reject-errors.ts
Comment (JoshuaKGoldberg) [reply packages/eslint-plugin/src/rules/prefer-promise-reject-errors.ts]: Ah splitting to variables in this way and then having the `if` _after_ them is a performance papercut. I'll post a new thread.

File: packages/eslint-plugin/src/rules/prefer-promise-reject-errors.ts
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/prefer-promise-reject-errors.ts:99]: [Performance] Previously, when it was all in one `if` statement, the type-checking functions `isPromiseConstructorLike` and `isPromiseLike` wouldn't be called if the previous logical bits (what is now `rejectMethodCalled`) were falsy.

So now it's a bit more readable at the cost of performance. One step forward, one step back!

My intent was to suggest splitting into functions so the logic gets lazily evaluated:

```ts
if (!rejectMethodCalled(callee) || !calleeIsLikePromise(callee)
```

...where `calleeObjectType` is only created inside `calleeIsLikePromise`. That way we don't call to the type checker at all if we don't need. The type checker can be slow sometimes so this is very good.

Aside: I filed an issue on myself to find a lint rule to enforce this: https://github.com/JoshuaKGoldberg/create-typescript-app/issues/1188

File: packages/eslint-plugin/tests/rules/prefer-promise-reject-errors.test.ts
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/tests/rules/prefer-promise-reject-errors.test.ts:1]: [Testing] Sorry for the late request, just realized a few missing cases:

* `Error[]`
* Readonly<Error>`
* `[Error]` (a tuple)
* `never`
* `unknown`

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
Sorry, two blocking requests:

* The variable readability change actually introduced a performance regression - lmk if the suggestion doesn't make sense?
* Apologies, I missed commenting on a couple of missing test cases

Since I already approved I do feel bad now requesting changes. If you don't have energy+time then no worries. I can apply these soon.

File: packages/eslint-plugin/src/rules/prefer-promise-reject-errors.ts
Comment (auvred) [reply packages/eslint-plugin/src/rules/prefer-promise-reject-errors.ts]: Ohh, don't know how I didn't notice that, thanks! Also thx for detailed explanation, I changed it https://github.com/typescript-eslint/typescript-eslint/pull/8011/commits/f186084686347b0f3850bf8cbb1ad07818e267c2

File: packages/eslint-plugin/tests/rules/prefer-promise-reject-errors.test.ts
Comment (auvred) [reply packages/eslint-plugin/tests/rules/prefer-promise-reject-errors.test.ts:1]: Nice catch, thanks! https://github.com/typescript-eslint/typescript-eslint/pull/8011/commits/e4af6eda213478034efc29276ae4b62537c1b16f

---

> * Readonly`

_I hope it's `ReadonlyArray<Error>`, right?_

Comment (JoshuaKGoldberg) [reply packages/eslint-plugin/tests/rules/prefer-promise-reject-errors.test.ts:1]: Oop, `Readonly<Error>`

Comment (auvred) [reply packages/eslint-plugin/tests/rules/prefer-promise-reject-errors.test.ts:1]: It was a nice try haha, I will add it tomorrow :)

Comment (auvred) [reply packages/eslint-plugin/tests/rules/prefer-promise-reject-errors.test.ts:1]: `Readonly<Error>` - is actually a great case!

I covered it in https://github.com/typescript-eslint/typescript-eslint/pull/8011/commits/232642964a8677e133ab283b15f096c8050763f8

---

TODO: after merging this PR, we should file an issue for `no-throw-literal` as it doesn't currently cover this case [playground](https://typescript-eslint.io/play/#ts=5.3.3&fileType=.tsx&code=GYVwdgxgLglg9mABMOcAUBTATlgXIgJQwEMATBAGwE8AeAURziwD4BKRAbwChFEoALLHADuibFi4BfIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1vwAtoOA7rUb5E0AIbx0UMf2iRwYAL4glQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

**Review by JoshuaKGoldberg — COMMENTED** [review]
Super, thanks! This looks great. 🚀 

Neil Patrick Harris giving two thumbs up and grinning

**Review by JoshuaKGoldberg — APPROVED** [review]

**Pull request merged**
